### PR TITLE
[tree] Reduce string allocations in TBranch*::GetFullName.

### PR DESCRIPTION
**Pull request by hageboeck**
When called repeatedly via TTree::GetBranch, the string allocations in
GetFullName have a notable runtime cost.
Here, the allocation is reduced to a single one in every branch.

See here for a performance analysis where repeated invocations produced significant overhead.
image

This seems to matter most if the branches have mothers, because in this case, the string composition kicks in.

@pcanal
- Do you know if there's a test that covers this case?
- Do you agree that this brings down TString allocations?

### CODE REVIEW

File: tree/tree/src/TBranchElement.cxx [hunk 1]
```diff
@@ -2845,11 +2845,19 @@ TString TBranchElement::GetFullName() const
       // The parent's name is already included in the name for split TClonesArray and STL collections
       return fName;
    }
-   TString motherName(mother->GetName());
-   if (motherName.Length() && (motherName[motherName.Length()-1] == '.')) {
+
```
Comment (Axel-Naumann) [inline tree/tree/src/TBranchElement.cxx:2848]: Can this just call `TBranch::GetFullName` at this point?

Comment (hageboeck) [reply tree/tree/src/TBranchElement.cxx:2848]: Even: Replace it by TBranch::GetFullName.
I propose to remove the override

Comment (hageboeck) [reply tree/tree/src/TBranchElement.cxx]: > I propose to remove the override

Done now.

**Review by pcanal — APPROVED** [review]

**Comment by pcanal** [comment]
thanks.

**Pull request merged**